### ISSUE TITLE
merge_and_unload returns the base (prior to fine-tuning) back!!!!

### ISSUE THREAD

**Issue by manitadayon** [huggingface/peft#2764]
I have fine-tune a model using PEFT and now I want to merge the base model to adapter. This is what I am doing:

```
base_model = AutoModelForCausalLM(model_id, device_map = 'auto')

model_finetuned  = PeftModel.from_pretrained(base_model, adapter_path)

```
Now the size of `model_finetuned `is roughly 42GB but when I do the following to merge the adapter into base:

`merged_model = model_finetuned.()
`
the size of `merged_model `is 36GB and its performance is like the base model, seems the adapter effect is gone.

I remember I used this feature in the past to get merged model, is anything changed? 

This is related post, where the last comment says this is normal, can someone elaborate?

https://github.com/huggingface/peft/issues/1836

Can I just save the `model_finetuned  ` as my merged model, can someone explain what is going on and why the merge_and_unload() is doing opposite of what it is supposed to do.

**Comment by BenjaminBossan** [comment]
> Now the size of `model_finetuned `is roughly 42GB but when I do the following to merge the adapter into base:
> 
> `merged_model = model_finetuned.() ` the size of `merged_model `is 36GB

What is the size of the base model before loading the PEFT adapter with `PeftModel.from_pretrained`? If it's 36GB, then everything is as expected. The model you get back after calling `merge_and_unload()` should be exactly the same architecture (layers, weights, etc.) as the original base model.

> its performance is like the base model, seems the adapter effect is gone.

This should of course not happen. How did you measure this? Could you please run inference _before_ calling `merge_and_unload()`, and then again after, and compare the results? Are they same or at least, very close (especially with quantized weights, there will be some deviation)? Or is the result identical to the base model before loading the adapter?

**Comment by manitadayon** [comment]
This is the issue, the result is identical to the base and not fine-tuned model. It is as I described the merge_and_unload() takes the fine tune model and give me the original base model. Like the impact of adapter is gone. Similar result before loading the adapter.

**Comment by BenjaminBossan** [comment]
Could you please show the code of how you run these tests?

**Comment by manitadayon** [comment]
Sure.
This is the process to get fine-tune model:

```
base_model = AutoModelForCausalLM(model_id, device_map = 'auto')

model_finetuned  = PeftModel.from_pretrained(base_model, adapter_path, adapter_name= 'A')
```

Then I make a inference on a question using hugging face transformers library:

This is pretty good result.
Not I want to merge the adapter to base, so here is what I do:

`Merged_model = model_finetuned.merge_and_unload()`

Now I do the same inference on the Merged_model
Which gives me response close to doing inference on the base model itself.

**Comment by BenjaminBossan** [comment]
Could you please provide the full code, including how you compare the outputs? Details are important here. Also, could you please share what the base model is? AFAICT, there is no quantization being applied, is that correct?

**Comment by manitadayon** [comment]
The model being quantized using `bitsandBytes `to 4 bits. Sure, here is how I do it for quick inference, the base model was Llama 70B 3.3. 

```
Inputs = tokenizer(prompt, return_tensors = 'pt')
output = model.generate(**Inputs.to('cuda'), max_new_tokens = 500)
tokenizer.decode(output[0], skip_special_tokens = True) 
```

**Comment by BenjaminBossan** [comment]
Okay, so as you're using 4bit bnb quantization, you should expect deviations after merging, as we're in a low precision regime. You should even see a warning:

> UserWarning: Merge lora module to 4-bit linear may get different generations due to rounding errors.

Moreover, when you want to compare the models, you shouldn't just use generation output. Generations are hard to compare and small differences can accumulate over time. Instead, you should compare continuous values like hidden states or logits.

Below I have a script that shows how to do this (for demonstration purposes, using a small Llama model):

```python
import os

import torch
from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig
from peft import LoraConfig, PeftModel, get_peft_model

model_id = "meta-llama/Llama-3.2-3B"
path = "/tmp/peft/2764"
device = 0
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_use_double_quant=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.bfloat16,
)

if not os.path.exists(path + "/adapter_model.safetensors"):
    # create a random LoRA checkpoint as stand in for a trained LoRA model
    model = AutoModelForCausalLM.from_pretrained(
        model_id, quantization_config=bnb_config, device_map=device
    )
    config = LoraConfig(target_modules=["q_proj", "v_proj"], init_lora_weights=False)
    model = get_peft_model(model, config)
    model.save_pretrained(path)
    del model
    torch.cuda.empty_cache()

tokenizer = AutoTokenizer.from_pretrained(model_id)
model = AutoModelForCausalLM.from_pretrained(
    model_id, quantization_config=bnb_config, device_map=device
)
inputs = tokenizer(["Hello, my dog is cute"], return_tensors="pt").to(model.device)
with torch.inference_mode():
    output_base = model(**inputs).logits

model = PeftModel.from_pretrained(model, path)
with torch.inference_mode():
    output_lora = model(**inputs).logits

merged = model.merge_and_unload()  # <= important: assign the output
with torch.inference_mode():
    output_merged = merged(**inputs).logits

mae_base = torch.mean(torch.abs(output_base - output_merged)).item()
mse_base = torch.mean((output_base - output_merged) ** 2).item()
mae_lora = torch.mean(torch.abs(output_merged - output_lora)).item()
mse_lora = torch.mean((output_merged - output_lora) ** 2).item()

print(f"MAE merged vs base: {mae_base:.6f}")
print(f"MSE merged vs base: {mse_base:.6f}")
print(f"MAE merged vs LoRA: {mae_lora:.6f}")
print(f"MSE merged vs LoRA: {mse_lora:.6f}")
```

For me, it prints:

```
MAE merged vs base: 0.413086
MSE merged vs base: 0.315674
MAE merged vs LoRA: 0.098450
MSE merged vs LoRA: 0.017105
```

So merging does add an error to the output, but the logits are still much closer to the logits from LoRA before merging than to the base model.

If you want to further reduce the error due to merging, check the [discussion in this thread](https://github.com/huggingface/peft/issues/2321).

**Comment by manitadayon** [comment]
I disagree with you on comparison of models. There is a reason we are fine tuning the model, and that's for my case for question answering task in my domains. The base model cannot answer any question but the fine-tune model can and it is quit good at it. 
Now when I merge the adapter then merged model is like the base model in every sense, even the response is very similar, it seems the impact of adapter is completely gone. This can only be seen using generation and to be honest this is what it matters.
Now if we have a couple of fine tune models we can use the error analysis to see which one is better.

As for the warning, yes I get that too.

**Comment by BenjaminBossan** [comment]
I see your point that at the end, we're interested in generation capability. However, right now, we're looking for the source of the error. For this, we need to find out:

1. Is your finding "just" because of the quantization error which is inherent with low bit quantization.
2. Or is this some more fundamental issue, e.g. a bug in PEFT.

To figure this out, we have to run a comparison such as the one I showed above. Therefore, if we want to make any progress, please run this comparison.

As for the final task, I'm sure you have some kind of validation or test set you used during/after training to measure how well the training worked. You can run this on 1) the base model, 2) the trained LoRA model, and 3) the merged model, and report the scores you get there. But this is only the next step.

**Comment by manitadayon** [comment]
Ok I will run the above test but it has nothing to do with quantization. The whole adapter impact is gone, after tuning the merge and unload it seems the model is retuned to base like it was never trained in the beginning.
Wondering why we just don't save the model after `PeftModel. from_pretrained` as the merged mode.

**Comment by BenjaminBossan** [comment]
> but it has nothing to do with quantization

It could very well be the reason. Merging requires adding the float weights from PEFT (most likely float32) into the quantized weights from the base model (int4), which is by its very nature a lossy operation that will change the model output.

> Wondering why we just don't save the model after `PeftModel. from_pretrained` as the merged mode.

If `PeftModel.from_pretrained` would save the merged model, the checkpoint would be quite large (the size of the base model) and not small (the size of the LoRA weights). Also, users often don't want to merge the weights, e.g. because they want to load, and switch between, multiple LoRA adapters.
